pkg/build/builder/common.go

```diff
@@ -82,3 +89,107 @@ func updateBuildRevision(c client.BuildInterface, build *api.Build, sourceInfo *
 		glog.Warningf("An error occurred saving build revision: %v", err)
 	}
 }
+
+// randomBuildTag generates a random tag used for building images in such a way
+// that the built image can be referred to unambiguously even in the face of
+// concurrent builds of the same originalTag.
+func randomBuildTag(originalTag string) string {
+	repo, _ := parsers.ParseRepositoryTag(originalTag)
+	suffix := fmt.Sprintf("%08x", rand.Uint32())
+	maxRepoLen := reference.NameTotalLengthMax - len(suffix) - 1
+	if len(repo) > maxRepoLen {
+		repo = repo[:maxRepoLen]
+	}
+	return fmt.Sprintf("%s:%s", repo, suffix)
+}
+
+// runPostCommitHook uses the dockerClient to run the post commit hook of build
+// in a new container based on image. The strategyName is used to generate an
+// appropriate name for the container.
+func runPostCommitHook(dockerClient DockerClient, build *api.Build, strategyName, image string) error {
+	glog.Infof("Running post commit hook with image %s ...", image)
+	containerName := fmt.Sprintf("openshift_%s-build_%s_%s_post-commit_%08x", strategyName, build.Name, build.Namespace, rand.Uint32())
+	return runScriptInContainer(dockerClient, build.Spec.PostCommit.RunBash, image, containerName)
+}
+
+// runScriptInContainer runs a Bash script in an ephemeral Docker container
+// started off image, and returns an error if the script cannot be run or
+// returns a non-zero exit code.
+func runScriptInContainer(client DockerClient, script, image, containerName string) error {
+	limits, err := cgroupResourceLimits()
+	if err != nil {
+		return err
+	}
+	return dockerRun(client, docker.CreateContainerOptions{
+		Name: containerName,
+		Config: &docker.Config{
+			Image:      image,
+			Entrypoint: []string{"bash", "-c"},
+			Cmd:        []string{script},
+		},
+		HostConfig: &docker.HostConfig{
+			// Limit container's resource allocation.
+			CPUShares: limits.CPUShares,
+			CPUPeriod: limits.CPUPeriod,
+			CPUQuota:  limits.CPUQuota,
+			Memory:    limits.Memory,
+		},
+	}, docker.LogsOptions{
+		// Stream logs to stdout and stderr.
+		OutputStream: os.Stdout,
+		ErrorStream:  os.Stderr,
+		Follow:       true,
+		Stdout:       true,
+		Stderr:       true,
+	})
+}
+
+// cgroupResourceLimit represents limits used to constrain resources allocated
+// by containers.
+type cgroupResourceLimit struct {
+	CPUShares int64
+	CPUPeriod int64
+	CPUQuota  int64
+	Memory    int64
+}
+
+// cgroupResourceLimits reads CPU and memory limits from /sys/fs/cgroup.
+func cgroupResourceLimits() (cgroupResourceLimit, error) {
+	r := intFromFileReader{}
+	cpuShares := r.read("/sys/fs/cgroup/cpu,cpuacct/cpu.shares")
+	cpuPeriod := r.read("/sys/fs/cgroup/cpu,cpuacct/cpu.cfs_period_us")
+	cpuQuota := r.read("/sys/fs/cgroup/cpu,cpuacct/cpu.cfs_quota_us")
+	memory := r.read("/sys/fs/cgroup/memory/memory.limit_in_bytes")
+	return cgroupResourceLimit{
+		CPUShares: cpuShares,
+		CPUPeriod: cpuPeriod,
+		CPUQuota:  cpuQuota,
+		Memory:    memory,
+	}, r.err
```

Comment:
my code has both cases.  For some errors (overflow) we ignore the error and return maxint64 because that's "close enough" in terms of the intended limit.

in an underflow case, we return a value (because we can't *not* return a value, it's not a pointer so nil is not an option) and an error.  The value we return is basically undefined, the caller can't rely on it for anything because an error occurred.

I think @rhcarvalho's situation here is similar.  If you are returning an error, it doesn't matter what else you return, any caller who relies on it is breaking the contract.